### PR TITLE
[dev-launcher] Launch directly into previously opened project

### PR DESCRIPTION
**Pull request by gabrieldonadel**
# Why

Closes ENG-10041

# How

In order to allow users to launch directly into previously opened projects, this PR updates `startWithWindow` and `handleIntent` to try to launch from the most recent URL of the RecentlyOpenedAppsRegistry. If the bundle cannot be loaded, it falls back to the regular launcher.

This also refactors the existing dev-client and dev-launcher config plugins adding a new `tryToLaunchLastOpenedBundle` parameter to the Android and iOS plugin config types.
 
E.g. 

```json
 "plugins": [
      [
        "expo-dev-client",
        {
          "android": {
            "tryToLaunchLastOpenedBundle": true
          },
          "ios": {
            "tryToLaunchLastOpenedBundle": true
          }
        }
      ]
    ]
```

# Test Plan

Run dev-client through bare-expo on Android and iOS testing opening the app directly, with and without "Recent Apps" and opening through deep links 

# Checklist
 

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/DevLauncherController.kt [hunk 1]
```diff
@@ -222,6 +222,18 @@ class DevLauncherController private constructor() :
       }
 
     intent?.let {
+      val shouldTryToLaunchLastOpenedBundle = getMetadataValue(context, "DEV_CLIENT_TRY_TO_LAUNCH_LAST_BUNDLE").toBoolean()
+      val lastOpenedApp = recentlyOpedAppsRegistry.getMostRecentApp()
+      if(shouldTryToLaunchLastOpenedBundle && lastOpenedApp != null && intent.action == Intent.ACTION_MAIN) {
```
Comment (lukmccall) [inline packages/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/DevLauncherController.kt:227]:
```suggestion
      if (shouldTryToLaunchLastOpenedBundle && lastOpenedApp != null && intent.action == Intent.ACTION_MAIN) {
```

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherRecentlyOpenedAppsRegistry.kt [hunk 2]
```diff
@@ -88,6 +88,16 @@ class DevLauncherRecentlyOpenedAppsRegistry(context: Context) {
     return result
   }
 
+  fun getMostRecentApp(): DevLauncherAppEntry? {
+    val recentlyOpenedApps = getRecentlyOpenedApps()
+
+    return if (recentlyOpenedApps.isNotEmpty()) {
+        recentlyOpenedApps.maxByOrNull { it.timestamp }
+    } else {
+        null
```
Comment (lukmccall) [inline packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherRecentlyOpenedAppsRegistry.kt:97]:
```suggestion
      recentlyOpenedApps.maxByOrNull { it.timestamp }
    } else {
      null
```

File: packages/expo-dev-launcher/ios/EXDevLauncherController.m [hunk 3]
```diff
@@ -249,18 +250,38 @@ - (void)startWithWindow:(UIWindow *)window delegate:(id<EXDevLauncherControllerD
...
 
-  if (!launchOptions[UIApplicationLaunchOptionsURLKey]) {
-    [self navigateToLauncher];
-  } else {
-    // For deeplink launch, we need the keyWindow for expo-splash-screen to setup correctly.
-    [_window makeKeyWindow];
+  if (launchOptions[UIApplicationLaunchOptionsURLKey]) {
+      // For deeplink launch, we need the keyWindow for expo-splash-screen to setup correctly.
+      [_window makeKeyWindow];
+      return;
```
Comment (lukmccall) [inline packages/expo-dev-launcher/ios/EXDevLauncherController.m:256]:
```suggestion
    // For deeplink launch, we need the keyWindow for expo-splash-screen to setup correctly.
    [_window makeKeyWindow];
    return;
```

File: packages/expo-dev-launcher/ios/EXDevLauncherController.m [hunk 4]
```diff
@@ -249,18 +250,38 @@ - (void)startWithWindow:(UIWindow *)window delegate:(id<EXDevLauncherControllerD
...
+              typeof(self) self = weakSelf;
+              if (!self) {
+                  return;
+              }
+
+              [self navigateToLauncher];
+          });
+      }];
+      return;
+  }
```
Comment (lukmccall) [inline packages/expo-dev-launcher/ios/EXDevLauncherController.m:273]:
```suggestion
  if (_lastOpenedAppUrl != nil && shouldTryToLaunchLastOpenedBundle) {
    [self loadApp:_lastOpenedAppUrl withProjectUrl:nil onSuccess:nil onError:^(NSError *error) {
       __weak typeof(self) weakSelf = self;
       dispatch_async(dispatch_get_main_queue(), ^{
         typeof(self) self = weakSelf;
         if (!self) {
           return;
         } 

         [self navigateToLauncher];
       });
    }];
    return;
  }
```

File: packages/expo-dev-launcher/ios/EXDevLauncherController.m [hunk 5]
```diff
@@ -249,18 +250,38 @@ - (void)startWithWindow:(UIWindow *)window delegate:(id<EXDevLauncherControllerD
...
+  [self navigateToLauncher];
 }
 
 - (void)autoSetupPrepare:(id<EXDevLauncherControllerDelegate>)delegate launchOptions:(NSDictionary * _Nullable)launchOptions
 {
   _delegate = delegate;
   _launchOptions = launchOptions;
+  NSDictionary *lastOpenedApp = [self.recentlyOpenedAppsRegistry mostRecentApp];
+  if(lastOpenedApp != nil){
```
Comment (lukmccall) [inline packages/expo-dev-launcher/ios/EXDevLauncherController.m:282]:
```suggestion
  if (lastOpenedApp != nil) {
```

**Review by lukmccall — COMMENTED** [review]
Can we add an option to set `tryToLaunchLastOpenedBundle` for both platforms without typing it twice?  Something like: 
```
 "plugins": [
      [
        "expo-dev-client",
        {
         	"tryToLaunchLastOpenedBundle": true
         }
      ]
    ]
```

**Comment by gabrieldonadel** [comment]
> Can we add an option to set `tryToLaunchLastOpenedBundle` for both platforms without typing it twice? Something like:
> 
> ```
>  "plugins": [
>       [
>         "expo-dev-client",
>         {
>          	"tryToLaunchLastOpenedBundle": true
>          }
>       ]
>     ]
> ```

Sure, updated

**Review by lukmccall — APPROVED** [review]

File: packages/expo-dev-launcher/plugin/src/pluginConfig.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Type representing available configuration for dev launcher.
+ */
+export type PluginConfigOptions = {
+  /**
+   * Attempts to launch directly into a previously opened project. If unable to connect,
+   * fall back to the launcher screen.
+   */
+  tryToLaunchLastOpenedBundle?: boolean;
```
Comment (brentvatne) [inline packages/expo-dev-launcher/plugin/src/pluginConfig.ts:32]: `tryToLaunchLastOpenedBundle` ->  `launchModeExperimental?: 'most-recent' | 'launcher'`

maybe? also fine with just adding an `Experimental` suffix

Comment (gabrieldonadel) [reply packages/expo-dev-launcher/plugin/src/pluginConfig.ts:32]: Sure, updated it here #24758

**Pull request merged**